cpp/src/parquet/column_reader_test.cc

```diff
@@ -757,6 +760,63 @@ TEST_F(RecordReaderTest, BasicReadRepeatedField) {
              /*levels_position=*/0);
 }
 
+// Tests reading the nullable values without saving a space for nulls.
+TEST_F(RecordReaderTest, ReadDenseForNullable) {
+  Init(/*max_def_level=*/1, /*max_rep_level=*/1, Repetition::REPEATED,
+       /*read_dense_for_nullable=*/true);
+
+  // Records look like: {[10], null, [20, 20], null, [30, 30, 30], null}
+  std::vector<std::shared_ptr<Page>> pages;
+  std::vector<int32_t> values = {10, 20, 20, 30, 30, 30};
+  std::vector<int16_t> def_levels = {1, 0, 1, 1, 0, 1, 1, 1, 0};
```

Comment:
we should probably add a test for when 3 level encoding lists so we can check behavior for a list containing a null element and an empty list?

Reply:
Done. Thanks for the suggestion, helped catch a bug reading Optional fields. I refactored the code a bit to make it less error-prone.